File: 0096_Unique Binary Search Trees/UniqueBinarySearchTrees.java
class Solution {
    private Map<Integer,Integer> map = new HashMap();
    public int numTrees(int n) {
        if(n <= 1) return 1;
        if(map.containsKey(n)) return map.get(n);
        
        int ans = 0;
        for(int root=1;root<=n;root++) {
            int left = numTrees(root - 1);
            int right = numTrees(n - root);
            ans += left * right;
        }
        
        map.put(n, ans);
        return ans;
    }
}

File: 0096_Unique Binary Search Trees/UniqueBinarySearchTrees.cs
public class Solution {
       
    public int NumTrees(int n) {
        var memo = new int[n+1];
        return NumTrees(n, memo);
    }
    
    private int NumTrees(int n, int[] memo)
    {
        if(n <= 1) return 1;
        if(memo[n] > 0) return memo[n];
        
        var sum = 0;
        // the node as a root is from to n inclusively
        for(int root=1;root<=n;root++)
        {
            //calculate the left nodes of root 
            var left = NumTrees(root-1, memo);
            //calculate the right nodes of root
            //make sure the total left + right should be equal to n - 1 (root - 1 + n - root == n-1)
            var right = NumTrees(n - root, memo);
            sum += left * right;
            
        }
        
        memo[n] = sum;
        return sum;
    }
}